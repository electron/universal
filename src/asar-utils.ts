import asar, { FileProperties } from '@electron/asar';
import { execFileSync } from 'child_process';
import crypto from 'crypto';
import fs from 'fs-extra';
import path from 'path';
import { minimatch } from 'minimatch';
import os from 'os';
import { d } from './debug';

const LIPO = 'lipo';

export enum AsarMode {
  NO_ASAR,
  HAS_ASAR,
}

export type MergeASARsOptions = {
  x64AsarPath: string;
  arm64AsarPath: string;
  outputAsarPath: string;

  singleArchFiles?: string;
};

// See: https://github.com/apple-opensource-mirror/llvmCore/blob/0c60489d96c87140db9a6a14c6e82b15f5e5d252/include/llvm/Object/MachOFormat.h#L108-L112
const MACHO_MAGIC = new Set([
  // 32-bit Mach-O
  0xfeedface, 0xcefaedfe,

  // 64-bit Mach-O
  0xfeedfacf, 0xcffaedfe,
]);

const MACHO_UNIVERSAL_MAGIC = new Set([
  // universal
  0xcafebabe, 0xbebafeca,
]);

export const detectAsarMode = async (appPath: string) => {
  d('checking asar mode of', appPath);
  const asarPath = path.resolve(appPath, 'Contents', 'Resources', 'app.asar');

  if (!(await fs.pathExists(asarPath))) {
    d('determined no asar');
    return AsarMode.NO_ASAR;
  }

  d('determined has asar');
  return AsarMode.HAS_ASAR;
};

export const generateAsarIntegrity = (asarPath: string) => {
  return {
    algorithm: 'SHA256' as const,
    hash: crypto
      .createHash('SHA256')
      .update(asar.getRawHeader(asarPath).headerString)
      .digest('hex'),
  };
};

export function toRelativePath(file: string): string {
  return file.replace(/^\//, '');
}

function isDirectory(a: string, file: string): boolean {
  return Boolean('files' in asar.statFile(a, file));
}

function checkSingleArch(archive: string, file: string, allowList?: string): void {
  if (allowList === undefined || !minimatch(file, allowList, { matchBase: true })) {
    throw new Error(
      `Detected unique file "${file}" in "${archive}" not covered by ` +
        `allowList rule: "${allowList}"`,
    );
  }
}

export const mergeASARs = async ({
  x64AsarPath,
  arm64AsarPath,
  outputAsarPath,
  singleArchFiles,
}: MergeASARsOptions): Promise<void> => {
  d(`merging ${x64AsarPath} and ${arm64AsarPath}`);

  const x64Dir = await fs.mkdtemp(path.join(os.tmpdir(), 'x64-'));
  const arm64Dir = await fs.mkdtemp(path.join(os.tmpdir(), 'arm64-'));

  const x64Files = new Set(asar.listPackage(x64AsarPath, { isPack: false }).map(toRelativePath));
  const arm64Files = new Set(
    asar.listPackage(arm64AsarPath, { isPack: false }).map(toRelativePath),
  );

  //
  // Build list of files/directories unique to each asar
  //

  for (const file of x64Files) {
    if (!arm64Files.has(file)) {
      checkSingleArch(x64AsarPath, file, singleArchFiles);
    }
  }
  const arm64Unique = [];
  for (const file of arm64Files) {
    if (!x64Files.has(file)) {
      checkSingleArch(arm64AsarPath, file, singleArchFiles);
      arm64Unique.push(file);
    }
  }

  //
  // Build set of unpacked directories and files and output as explicit ordering file
  // Note: We don't know the ordering of unique files, so we exclude them from the ordering file?
  //

  const ordering = await generateOrderingConfig(
    { asarPath: x64AsarPath, files: x64Files },
    { asarPath: arm64AsarPath, files: arm64Files },
  );

  //
  // Find common bindings with different content
  //

  const commonBindings = [];
  for (const file of x64Files) {
    if (!arm64Files.has(file)) {
      continue;
    }

    // Skip directories
    if (isDirectory(x64AsarPath, file)) {
      continue;
    }

    const x64Content = asar.extractFile(x64AsarPath, file);
    const arm64Content = asar.extractFile(arm64AsarPath, file);

    if (x64Content.compare(arm64Content) === 0) {
      continue;
    }

    if (
      MACHO_UNIVERSAL_MAGIC.has(x64Content.readUInt32LE(0)) &&
      MACHO_UNIVERSAL_MAGIC.has(arm64Content.readUInt32LE(0))
    ) {
      continue;
    }

    if (!MACHO_MAGIC.has(x64Content.readUInt32LE(0))) {
      throw new Error(`Can't reconcile two non-macho files ${file}`);
    }

    commonBindings.push(file);
  }

  //
  // Extract both
  //

  try {
    d(`extracting ${x64AsarPath} to ${x64Dir}`);
    asar.extractAll(x64AsarPath, x64Dir);

    d(`extracting ${arm64AsarPath} to ${arm64Dir}`);
    asar.extractAll(arm64AsarPath, arm64Dir);

    for (const file of arm64Unique) {
      const source = path.resolve(arm64Dir, file);
      const destination = path.resolve(x64Dir, file);

      if (isDirectory(arm64AsarPath, file)) {
        d(`creating unique directory: ${file}`);
        await fs.mkdirp(destination);
        continue;
      }

      d(`xopying unique file: ${file}`);
      await fs.mkdirp(path.dirname(destination));
      await fs.copy(source, destination);
    }

    for (const binding of commonBindings) {
      const source = await fs.realpath(path.resolve(arm64Dir, binding));
      const destination = await fs.realpath(path.resolve(x64Dir, binding));

      d(`merging binding: ${binding}`);
      execFileSync(LIPO, [source, destination, '-create', '-output', destination]);
    }

    d(`creating archive at ${outputAsarPath}`);

    await asar.createPackageWithOptions(x64Dir, outputAsarPath, { ordering });

    d('done merging');
  } finally {
    await Promise.all([fs.remove(x64Dir), fs.remove(arm64Dir)]);
  }
};

type ArchSpecificOptions = {
  asarPath: string;
  files: Set<string>;
};

export async function generateOrderingConfig(x64: ArchSpecificOptions, arm64: ArchSpecificOptions) {
  const ordering: Record<string, FileProperties['properties']> = {};

  function buildUnpacked(asarPath: string, fileList: Set<string>): void {
    for (const file of fileList) {
      const stat = asar.statFile(asarPath, file);

      let unpack = 'unpacked' in stat && stat.unpacked;

      if ('files' in stat) {
        continue;
      }

      ordering[file] = { unpack };
    }
  }

  buildUnpacked(x64.asarPath, x64.files);
  buildUnpacked(arm64.asarPath, arm64.files);

  const data = Object.entries(ordering).reduce((prev, curr) => {
    return `${prev}${curr[0]}:${JSON.stringify(curr[1])}\n`;
  }, '');
  const orderingFile = path.join(os.tmpdir(), 'ordering.txt');
  await fs.writeFile(orderingFile, data);
  return orderingFile;
}
