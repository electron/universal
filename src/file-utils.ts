import { spawn, ExitCodeError } from '@malept/cross-spawn-promise';
import * as fs from 'fs-extra';
import * as path from 'path';

const MACHO_PREFIX = 'Mach-O ';

export enum AppFileType {
  MACHO,
  PLAIN,
  SNAPSHOT,
  APP_CODE,
}

export type AppFile = {
  relativePath: string;
  type: AppFileType;
};

/**
 *
 * @param appPath Path to the application
 */
export const getAllAppFiles = async (appPath: string): Promise<AppFile[]> => {
  const files: AppFile[] = [];

  const visited = new Set<string>();
  const traverse = async (p: string) => {
    p = await fs.realpath(p);
    if (visited.has(p)) return;
    visited.add(p);

    const info = await fs.stat(p);
    if (info.isSymbolicLink()) return;
    if (info.isFile()) {
      let fileType = AppFileType.PLAIN;

      var fileOutput = '';
      try {
        fileOutput = await spawn('file', ['--brief', '--no-pad', p]);
      } catch (e) {
        if (e instanceof ExitCodeError) {
          /* silently accept error codes from "file" */
        } else {
          throw e;
        }
      }
      if (p.includes('app.asar')) {
        fileType = AppFileType.APP_CODE;
      } else if (fileOutput.startsWith(MACHO_PREFIX)) {
        fileType = AppFileType.MACHO;
      } else if (p.endsWith('.bin')) {
        fileType = AppFileType.SNAPSHOT;
      }

      files.push({
        relativePath: path.relative(appPath, p),
        type: fileType,
      });
    }

    if (info.isDirectory()) {
      for (const child of await fs.readdir(p)) {
        await traverse(path.resolve(p, child));
      }
    }
  };
  await traverse(appPath);

  return files;
};
