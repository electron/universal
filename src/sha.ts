import * as fs from 'fs-extra';
import * as crypto from 'crypto';
import { pipeline } from 'stream/promises';

import { d } from './debug';

export const sha = async (filePath: string) => {
  d('hashing', filePath);
  const hash = crypto.createHash('sha256');
  hash.setEncoding('hex');
  await pipeline(fs.createReadStream(filePath), hash);
  return hash.read();
};
