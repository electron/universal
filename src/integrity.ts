import fs from 'node:fs';
import path from 'node:path';

import { AppFileType, getAllAppFiles } from './file-utils.js';
import { generateAsarIntegrity } from './asar-utils.js';

type IntegrityMap = {
  [filepath: string]: string;
};

export interface HeaderHash {
  algorithm: 'SHA256';
  hash: string;
}

export interface AsarIntegrity {
  [key: string]: HeaderHash;
}

export async function computeIntegrityData(contentsPath: string): Promise<AsarIntegrity> {
  const root = await fs.promises.realpath(contentsPath);

  const resourcesRelativePath = 'Resources';
  const resourcesPath = path.resolve(root, resourcesRelativePath);

  const resources = await getAllAppFiles(resourcesPath);
  const resourceAsars = resources
    .filter((file) => file.type === AppFileType.APP_CODE)
    .reduce<IntegrityMap>(
      (prev, file) => ({
        ...prev,
        [path.join(resourcesRelativePath, file.relativePath)]: path.join(
          resourcesPath,
          file.relativePath,
        ),
      }),
      {},
    );

  // sort to produce constant result
  const allAsars = Object.entries(resourceAsars).sort(([name1], [name2]) =>
    name1.localeCompare(name2),
  );
  const hashes = await Promise.all(allAsars.map(async ([, from]) => generateAsarIntegrity(from)));
  const asarIntegrity: AsarIntegrity = {};
  for (let i = 0; i < allAsars.length; i++) {
    const [asar] = allAsars[i];
    asarIntegrity[asar] = hashes[i];
  }
  return asarIntegrity;
}
