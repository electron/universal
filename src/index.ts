import { spawn } from '@malept/cross-spawn-promise';
import * as asar from '@electron/asar';
import * as fs from 'fs-extra';
import { minimatch } from 'minimatch';
import * as os from 'os';
import * as path from 'path';
import * as plist from 'plist';
import * as dircompare from 'dir-compare';

import { AppFile, AppFileType, getAllAppFiles } from './file-utils';
import { AsarMode, detectAsarMode, generateAsarIntegrity, mergeASARs } from './asar-utils';
import { sha } from './sha';
import { d } from './debug';

/**
 * Options to pass into the {@link makeUniversalApp} function.
 *
 * Requires absolute paths for input x64 and arm64 apps and an absolute path to the
 * output universal app.
 */
export type MakeUniversalOpts = {
  /**
   * Absolute file system path to the x64 version of your application (e.g. `/Foo/bar/MyApp_x64.app`).
   */
  x64AppPath: string;
  /**
   * Absolute file system path to the arm64 version of your application (e.g. `/Foo/bar/MyApp_arm64.app`).
   */
  arm64AppPath: string;
  /**
   * Absolute file system path you want the universal app to be written to (e.g. `/Foo/var/MyApp_universal.app`).
   *
   * If this file exists on disk already, it will be overwritten ONLY if {@link MakeUniversalOpts.force} is set to `true`.
   */
  outAppPath: string;
  /**
   * Forcefully overwrite any existing files that are in the way of generating the universal application.
   *
   * @defaultValue `false`
   */
  force?: boolean;
  /**
   * Merge x64 and arm64 ASARs into one.
   *
   * @defaultValue `false`
   */
  mergeASARs?: boolean;
  /**
   * If {@link MakeUniversalOpts.mergeASARs} is enabled, this property provides a
   * {@link https://github.com/isaacs/minimatch?tab=readme-ov-file#features | minimatch}
   * pattern of paths that are allowed to be present in one of the ASAR files, but not in the other.
   *
   */
  singleArchFiles?: string;
  /**
   * A {@link https://github.com/isaacs/minimatch?tab=readme-ov-file#features | minimatch}
   * pattern of binaries that are expected to be the same x64 binary in both of the application files.
   *
   * Use this if your application contains binaries that have already been merged into a universal file
   * using the `lipo` tool.
   *
   * @see Apple's {@link https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary | Building a universal macOS binary} documentation
   */
  x64ArchFiles?: string;
  /**
   * A {@link https://github.com/isaacs/minimatch?tab=readme-ov-file#features | minimatch} pattern of `Info.plist`
   * paths that should not receive an injected `ElectronAsarIntegrity` value.
   *
   * Use this if your application contains another bundle that's already signed.
   */
  infoPlistsToIgnore?: string;
};

const dupedFiles = (files: AppFile[]) =>
  files.filter((f) => f.type !== AppFileType.SNAPSHOT && f.type !== AppFileType.APP_CODE);

export const makeUniversalApp = async (opts: MakeUniversalOpts): Promise<void> => {
  d('making a universal app with options', opts);

  if (process.platform !== 'darwin')
    throw new Error('@electron/universal is only supported on darwin platforms');
  if (!opts.x64AppPath || !path.isAbsolute(opts.x64AppPath))
    throw new Error('Expected opts.x64AppPath to be an absolute path but it was not');
  if (!opts.arm64AppPath || !path.isAbsolute(opts.arm64AppPath))
    throw new Error('Expected opts.arm64AppPath to be an absolute path but it was not');
  if (!opts.outAppPath || !path.isAbsolute(opts.outAppPath))
    throw new Error('Expected opts.outAppPath to be an absolute path but it was not');

  if (await fs.pathExists(opts.outAppPath)) {
    d('output path exists already');
    if (!opts.force) {
      throw new Error(
        `The out path "${opts.outAppPath}" already exists and force is not set to true`,
      );
    } else {
      d('overwriting existing application because force == true');
      await fs.remove(opts.outAppPath);
    }
  }

  const x64AsarMode = await detectAsarMode(opts.x64AppPath);
  const arm64AsarMode = await detectAsarMode(opts.arm64AppPath);
  d('detected x64AsarMode =', x64AsarMode);
  d('detected arm64AsarMode =', arm64AsarMode);

  if (x64AsarMode !== arm64AsarMode)
    throw new Error(
      'Both the x64 and arm64 versions of your application need to have been built with the same asar settings (enabled vs disabled)',
    );

  const tmpDir = await fs.mkdtemp(path.resolve(os.tmpdir(), 'electron-universal-'));
  d('building universal app in', tmpDir);

  try {
    d('copying x64 app as starter template');
    const tmpApp = path.resolve(tmpDir, 'Tmp.app');
    await spawn('cp', ['-R', opts.x64AppPath, tmpApp]);

    const uniqueToX64: string[] = [];
    const uniqueToArm64: string[] = [];
    const x64Files = await getAllAppFiles(await fs.realpath(tmpApp));
    const arm64Files = await getAllAppFiles(await fs.realpath(opts.arm64AppPath));

    for (const file of dupedFiles(x64Files)) {
      if (!arm64Files.some((f) => f.relativePath === file.relativePath))
        uniqueToX64.push(file.relativePath);
    }
    for (const file of dupedFiles(arm64Files)) {
      if (!x64Files.some((f) => f.relativePath === file.relativePath))
        uniqueToArm64.push(file.relativePath);
    }
    if (uniqueToX64.length !== 0 || uniqueToArm64.length !== 0) {
      d('some files were not in both builds, aborting');
      console.error({
        uniqueToX64,
        uniqueToArm64,
      });
      throw new Error(
        'While trying to merge mach-o files across your apps we found a mismatch, the number of mach-o files is not the same between the arm64 and x64 builds',
      );
    }

    for (const file of x64Files.filter((f) => f.type === AppFileType.PLAIN)) {
      const x64Sha = await sha(path.resolve(opts.x64AppPath, file.relativePath));
      const arm64Sha = await sha(path.resolve(opts.arm64AppPath, file.relativePath));
      if (x64Sha !== arm64Sha) {
        d('SHA for file', file.relativePath, `does not match across builds ${x64Sha}!=${arm64Sha}`);
        // The MainMenu.nib files generated by Xcode13 are deterministic in effect but not deterministic in generated sequence
        if (path.basename(path.dirname(file.relativePath)) === 'MainMenu.nib') {
          // The mismatch here is OK so we just move on to the next one
          continue;
        }
        throw new Error(
          `Expected all non-binary files to have identical SHAs when creating a universal build but "${file.relativePath}" did not`,
        );
      }
    }
    const knownMergedMachOFiles = new Set();
    for (const machOFile of x64Files.filter((f) => f.type === AppFileType.MACHO)) {
      const first = await fs.realpath(path.resolve(tmpApp, machOFile.relativePath));
      const second = await fs.realpath(path.resolve(opts.arm64AppPath, machOFile.relativePath));

      const x64Sha = await sha(path.resolve(opts.x64AppPath, machOFile.relativePath));
      const arm64Sha = await sha(path.resolve(opts.arm64AppPath, machOFile.relativePath));
      if (x64Sha === arm64Sha) {
        if (
          opts.x64ArchFiles === undefined ||
          !minimatch(machOFile.relativePath, opts.x64ArchFiles, { matchBase: true })
        ) {
          throw new Error(
            `Detected file "${machOFile.relativePath}" that's the same in both x64 and arm64 builds and not covered by the ` +
              `x64ArchFiles rule: "${opts.x64ArchFiles}"`,
          );
        }

        d(
          'SHA for Mach-O file',
          machOFile.relativePath,
          `matches across builds ${x64Sha}===${arm64Sha}, skipping lipo`,
        );
        continue;
      }

      d('joining two MachO files with lipo', {
        first,
        second,
      });
      await spawn('lipo', [
        first,
        second,
        '-create',
        '-output',
        await fs.realpath(path.resolve(tmpApp, machOFile.relativePath)),
      ]);
      knownMergedMachOFiles.add(machOFile.relativePath);
    }

    /**
     * If we don't have an ASAR we need to check if the two "app" folders are identical, if
     * they are then we can just leave one there and call it a day.  If the app folders for x64
     * and arm64 are different though we need to rename each folder and create a new fake "app"
     * entrypoint to dynamically load the correct app folder
     */
    if (x64AsarMode === AsarMode.NO_ASAR) {
      d('checking if the x64 and arm64 app folders are identical');
      const comparison = await dircompare.compare(
        path.resolve(tmpApp, 'Contents', 'Resources', 'app'),
        path.resolve(opts.arm64AppPath, 'Contents', 'Resources', 'app'),
        { compareSize: true, compareContent: true },
      );
      const differences = comparison.diffSet!.filter((difference) => difference.state !== 'equal');
      d(`Found ${differences.length} difference(s) between the x64 and arm64 folders`);
      const nonMergedDifferences = differences.filter(
        (difference) =>
          !difference.name1 ||
          !knownMergedMachOFiles.has(
            path.join('Contents', 'Resources', 'app', difference.relativePath, difference.name1),
          ),
      );
      d(`After discluding MachO files merged with lipo ${nonMergedDifferences.length} remain.`);

      if (nonMergedDifferences.length > 0) {
        d('x64 and arm64 app folders are different, creating dynamic entry ASAR');
        await fs.move(
          path.resolve(tmpApp, 'Contents', 'Resources', 'app'),
          path.resolve(tmpApp, 'Contents', 'Resources', 'app-x64'),
        );
        await fs.copy(
          path.resolve(opts.arm64AppPath, 'Contents', 'Resources', 'app'),
          path.resolve(tmpApp, 'Contents', 'Resources', 'app-arm64'),
        );

        const entryAsar = path.resolve(tmpDir, 'entry-asar');
        await fs.mkdir(entryAsar);
        await fs.copy(
          path.resolve(__dirname, '..', '..', 'entry-asar', 'no-asar.js'),
          path.resolve(entryAsar, 'index.js'),
        );
        let pj = await fs.readJson(
          path.resolve(opts.x64AppPath, 'Contents', 'Resources', 'app', 'package.json'),
        );
        pj.main = 'index.js';
        await fs.writeJson(path.resolve(entryAsar, 'package.json'), pj);
        await asar.createPackage(
          entryAsar,
          path.resolve(tmpApp, 'Contents', 'Resources', 'app.asar'),
        );
      } else {
        d('x64 and arm64 app folders are the same');
      }
    }

    const generatedIntegrity: Record<string, { algorithm: 'SHA256'; hash: string }> = {};

    /**
     * If we have an ASAR we just need to check if the two "app.asar" files have the same hash,
     * if they are, same as above, we can leave one there and call it a day.  If they're different
     * we have to make a dynamic entrypoint.  There is an assumption made here that every file in
     * app.asar.unpacked is a native node module.  This assumption _may_ not be true so we should
     * look at codifying that assumption as actual logic.
     */
    // FIXME: Codify the assumption that app.asar.unpacked only contains native modules
    if (x64AsarMode === AsarMode.HAS_ASAR) {
      d('checking if the x64 and arm64 asars are identical');
      const x64AsarSha = await sha(path.resolve(tmpApp, 'Contents', 'Resources', 'app.asar'));
      const arm64AsarSha = await sha(
        path.resolve(opts.arm64AppPath, 'Contents', 'Resources', 'app.asar'),
      );

      if (x64AsarSha !== arm64AsarSha) {
        d('x64 and arm64 asars are different');
        if (opts.mergeASARs) {
          d('merging x64 and arm64 asars');
          const output = path.resolve(tmpApp, 'Contents', 'Resources', 'app.asar');
          await mergeASARs({
            x64AsarPath: path.resolve(tmpApp, 'Contents', 'Resources', 'app.asar'),
            arm64AsarPath: path.resolve(opts.arm64AppPath, 'Contents', 'Resources', 'app.asar'),
            outputAsarPath: output,
            singleArchFiles: opts.singleArchFiles,
          });

          generatedIntegrity['Resources/app.asar'] = generateAsarIntegrity(output);
        } else {
          d('splitting x64 and arm64 asars');
          const x64AsarPath = path.resolve(tmpApp, 'Contents', 'Resources', 'app-x64.asar');
          await fs.move(path.resolve(tmpApp, 'Contents', 'Resources', 'app.asar'), x64AsarPath);
          const x64Unpacked = path.resolve(tmpApp, 'Contents', 'Resources', 'app.asar.unpacked');
          if (await fs.pathExists(x64Unpacked)) {
            await fs.move(
              x64Unpacked,
              path.resolve(tmpApp, 'Contents', 'Resources', 'app-x64.asar.unpacked'),
            );
          }

          const arm64AsarPath = path.resolve(tmpApp, 'Contents', 'Resources', 'app-arm64.asar');
          await fs.copy(
            path.resolve(opts.arm64AppPath, 'Contents', 'Resources', 'app.asar'),
            arm64AsarPath,
          );
          const arm64Unpacked = path.resolve(
            opts.arm64AppPath,
            'Contents',
            'Resources',
            'app.asar.unpacked',
          );
          if (await fs.pathExists(arm64Unpacked)) {
            await fs.copy(
              arm64Unpacked,
              path.resolve(tmpApp, 'Contents', 'Resources', 'app-arm64.asar.unpacked'),
            );
          }

          const entryAsar = path.resolve(tmpDir, 'entry-asar');
          await fs.mkdir(entryAsar);
          await fs.copy(
            path.resolve(__dirname, '..', '..', 'entry-asar', 'has-asar.js'),
            path.resolve(entryAsar, 'index.js'),
          );
          let pj = JSON.parse(
            asar
              .extractFile(
                path.resolve(opts.x64AppPath, 'Contents', 'Resources', 'app.asar'),
                'package.json',
              )
              .toString('utf8'),
          );
          pj.main = 'index.js';
          await fs.writeJson(path.resolve(entryAsar, 'package.json'), pj);
          const asarPath = path.resolve(tmpApp, 'Contents', 'Resources', 'app.asar');
          await asar.createPackage(entryAsar, asarPath);

          generatedIntegrity['Resources/app.asar'] = generateAsarIntegrity(asarPath);
          generatedIntegrity['Resources/app-x64.asar'] = generateAsarIntegrity(x64AsarPath);
          generatedIntegrity['Resources/app-arm64.asar'] = generateAsarIntegrity(arm64AsarPath);
        }
      } else {
        d('x64 and arm64 asars are the same');
        generatedIntegrity['Resources/app.asar'] = generateAsarIntegrity(
          path.resolve(tmpApp, 'Contents', 'Resources', 'app.asar'),
        );
      }
    }

    const plistFiles = x64Files.filter((f) => f.type === AppFileType.INFO_PLIST);
    for (const plistFile of plistFiles) {
      const x64PlistPath = path.resolve(opts.x64AppPath, plistFile.relativePath);
      const arm64PlistPath = path.resolve(opts.arm64AppPath, plistFile.relativePath);

      const { ElectronAsarIntegrity: x64Integrity, ...x64Plist } = plist.parse(
        await fs.readFile(x64PlistPath, 'utf8'),
      ) as any;
      const { ElectronAsarIntegrity: arm64Integrity, ...arm64Plist } = plist.parse(
        await fs.readFile(arm64PlistPath, 'utf8'),
      ) as any;
      if (JSON.stringify(x64Plist) !== JSON.stringify(arm64Plist)) {
        throw new Error(
          `Expected all Info.plist files to be identical when ignoring integrity when creating a universal build but "${plistFile.relativePath}" was not`,
        );
      }

      const injectAsarIntegrity =
        !opts.infoPlistsToIgnore ||
        minimatch(plistFile.relativePath, opts.infoPlistsToIgnore, { matchBase: true });
      const mergedPlist = injectAsarIntegrity
        ? { ...x64Plist, ElectronAsarIntegrity: generatedIntegrity }
        : { ...x64Plist };

      await fs.writeFile(path.resolve(tmpApp, plistFile.relativePath), plist.build(mergedPlist));
    }

    for (const snapshotsFile of arm64Files.filter((f) => f.type === AppFileType.SNAPSHOT)) {
      d('copying snapshot file', snapshotsFile.relativePath, 'to target application');
      await fs.copy(
        path.resolve(opts.arm64AppPath, snapshotsFile.relativePath),
        path.resolve(tmpApp, snapshotsFile.relativePath),
      );
    }

    d('moving final universal app to target destination');
    await fs.mkdirp(path.dirname(opts.outAppPath));
    await spawn('mv', [tmpApp, opts.outAppPath]);
  } catch (err) {
    throw err;
  } finally {
    await fs.remove(tmpDir);
  }
};
