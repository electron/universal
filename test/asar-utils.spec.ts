import * as path from 'path';

import { AsarMode, detectAsarMode, generateAsarIntegrity } from '../src/asar-utils';

const asarsPath = path.resolve(__dirname, 'fixtures', 'asars');
const appsPath = path.resolve(__dirname, 'fixtures', 'apps');

describe('asar-utils', () => {
  describe('detectAsarMode', () => {
    it('should correctly detect an asar enabled app', async () => {
      expect(await detectAsarMode(path.resolve(appsPath, 'Arm64Asar.app'))).toBe(AsarMode.HAS_ASAR);
    });

    it('should correctly detect an app without an asar', async () => {
      expect(await detectAsarMode(path.resolve(appsPath, 'Arm64NoAsar.app'))).toBe(
        AsarMode.NO_ASAR,
      );
    });
  });

  describe('generateAsarIntegrity', () => {
    it('should deterministically hash an asar header', async () => {
      expect(generateAsarIntegrity(path.resolve(asarsPath, 'app.asar')).hash).toEqual(
        '85fff474383bd8df11cd9c5784e8fcd1525af71ff140a8a882e1dc9d5b39fcbf',
      );
    });
  });
});
