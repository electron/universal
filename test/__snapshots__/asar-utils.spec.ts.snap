// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asar-utils generateOrderingConfig should properly read \`unpacked\` flag and generate ordering config 1`] = `
"dir1/file1.txt:{"unpack":false}
dir2/file2.png:{"unpack":true}
dir2/file3.txt:{"unpack":true}
emptyfile.txt:{"unpack":false}
file0.txt:{"unpack":false}
"
`;
